File: mads/godot/client_libs/python/godot/loglib.py
import os
import sys
import json
import datetime
import logging
from typing import Optional, List, Dict, Any

import redis

# Define log levels
LEVELS = {"TRACE": 5, "DEBUG": 10, "INFO": 20, "WARN": 30, "ERROR": 40}
DEFAULT_REDACT_FIELDS = ['password', 'token', 'api_key', 'authorization', 'secret']
REDACTED_VALUE = "[REDACTED]"

class ICCMLogger:
    """Non-blocking logger that sends logs to Godot via Redis (REQ-LIB-003)"""

    def __init__(self,
                 component: str,
                 redis_url: str = "redis://localhost:6379",
                 default_level: str = "INFO",
                 redact_fields: Optional[List[str]] = None,
                 max_queue_size: int = 100000,
                 log_queue_name: str = "logs:queue"):

        self.component = component
        self.level = LEVELS.get(default_level.upper(), LEVELS["INFO"])
        self.redact_fields = set((redact_fields or []) + DEFAULT_REDACT_FIELDS)
        self.max_queue_size = max_queue_size
        self.log_queue_name = log_queue_name

        try:
            self.redis_client = redis.from_url(redis_url, socket_connect_timeout=1, decode_responses=True)
            self.redis_client.ping()
            self._redis_available = True
            self._load_lua_script()
        except (redis.exceptions.ConnectionError, redis.exceptions.TimeoutError) as e:
            self._redis_available = False
            self.redis_client = None
            self._log_fallback_warning(f"Could not connect to Redis at {redis_url}. Reason: {e}. Falling back to local logging.")

    def _load_lua_script(self):
        """Load atomic FIFO queue management script (REQ-GOD-005)"""
        # REQ-PERF-003: Enforce 100,000 entry limit with FIFO drop
        script = """
        local queue = KEYS[1]
        local max_size = tonumber(ARGV[1])
        local log_entry = ARGV[2]

        local current_size = redis.call('LLEN', queue)
        if current_size >= max_size then
            redis.call('RPOP', queue)
        end
        redis.call('LPUSH', queue, log_entry)
        return current_size + 1
        """
        self.script_sha = self.redis_client.script_load(script)

    def set_level(self, level: str):
        """REQ-LIB-005: Dynamically change the log level"""
        new_level = LEVELS.get(level.upper())
        if new_level is not None:
            self.level = new_level
        else:
            self._log_fallback_warning(f"Invalid log level '{level}'. Level not changed.")

    def _redact(self, data: Any) -> Any:
        """REQ-SEC-001/002/003: Recursively redact sensitive fields"""
        if not isinstance(data, dict):
            return data

        clean_data = {}
        for key, value in data.items():
            if isinstance(key, str) and key.lower() in self.redact_fields:
                clean_data[key] = REDACTED_VALUE
            elif isinstance(value, dict):
                clean_data[key] = self._redact(value)
            elif isinstance(value, list):
                clean_data[key] = [self._redact(item) for item in value]
            else:
                clean_data[key] = value
        return clean_data

    def _log(self, level_name: str, message: str, data: Optional[Dict] = None, trace_id: Optional[str] = None):
        level_num = LEVELS[level_name]
        if level_num < self.level:
            return  # REQ-LIB-006: Filter logs below current level

        log_entry = {
            "component": self.component,
            "level": level_name,
            "message": message,
            "trace_id": trace_id,  # REQ-COR-004, REQ-COR-005
            "data": self._redact(data) if data else None,
            "created_at": datetime.datetime.now(datetime.timezone.utc).isoformat()
        }

        # REQ-LIB-003: Non-blocking push with local fallback
        if not self._redis_available:
            self._log_local(log_entry)
            return

        try:
            # Use Lua script for atomic FIFO queue management
            self.redis_client.evalsha(
                self.script_sha,
                1,
                self.log_queue_name,
                self.max_queue_size,
                json.dumps(log_entry)
            )
        except (redis.exceptions.ConnectionError, redis.exceptions.TimeoutError) as e:
            # REQ-LIB-004: Fallback on ANY Redis failure
            self._redis_available = False
            self._log_fallback_warning(f"Redis connection failed. Reason: {e}. Falling back to local logging.")
            self._log_local(log_entry)

    def _log_fallback_warning(self, message: str):
        """REQ-LIB-007: Log a warning indicating fallback mode"""
        print(f"[ICCMLogger WARNING] {self.component}: {message}", file=sys.stderr)

    def _log_local(self, log_entry: Dict):
        """Logs to stdout as a fallback"""
        print(json.dumps(log_entry), file=sys.stdout)

    # REQ-LIB-001: Public log methods
    def trace(self, message: str, data: Optional[Dict] = None, trace_id: Optional[str] = None):
        self._log("TRACE", message, data, trace_id)

    def debug(self, message: str, data: Optional[Dict] = None, trace_id: Optional[str] = None):
        self._log("DEBUG", message, data, trace_id)

    def info(self, message: str, data: Optional[Dict] = None, trace_id: Optional[str] = None):
        self._log("INFO", message, data, trace_id)

    def warn(self, message: str, data: Optional[Dict] = None, trace_id: Optional[str] = None):
        self._log("WARN", message, data, trace_id)

    def error(self, message: str, data: Optional[Dict] = None, trace_id: Optional[str] = None):
        self._log("ERROR", message, data, trace_id)

    def close(self):
        """Gracefully close the Redis connection"""
        if self.redis_client:
            self.redis_client.close()

File: mads/gates/loglib.js
import redis from 'redis';

const LEVELS = { "TRACE": 5, "DEBUG": 10, "INFO": 20, "WARN": 30, "ERROR": 40 };
const DEFAULT_REDACT_FIELDS = new Set(['password', 'token', 'api_key', 'authorization', 'secret']);
const REDACTED_VALUE = "[REDACTED]";

export class ICCMLogger {
    /**
     * Non-blocking logger that sends logs to Godot via Redis (REQ-LIB-003)
     * @param {object} options
     * @param {string} options.component - Component name
     * @param {string} [options.redisUrl='redis://localhost:6379']
     * @param {string} [options.defaultLevel='INFO']
     * @param {string[]} [options.redactFields=[]]
     * @param {number} [options.maxQueueSize=100000]
     * @param {string} [options.logQueueName='logs:queue']
     */
    constructor({ component, redisUrl = 'redis://localhost:6379', defaultLevel = 'INFO', redactFields = [], maxQueueSize = 100000, logQueueName = 'logs:queue' }) {
        if (!component) {
            throw new Error('ICCMLogger requires a "component" name.');
        }
        this.component = component;
        this.level = LEVELS[defaultLevel.toUpperCase()] || LEVELS.INFO;
        this.redactFields = new Set([...DEFAULT_REDACT_FIELDS, ...redactFields]);
        this.maxQueueSize = maxQueueSize;
        this.logQueueName = logQueueName;
        this._redisAvailable = false;

        this.redisClient = redis.createClient({
            url: redisUrl,
            socket: { connectTimeout: 1000 }
        });

        // REQ-GOD-005: Lua script for atomic FIFO queue management
        this.luaScript = `
            local queue = KEYS[1]
            local max_size = tonumber(ARGV[1])
            local log_entry = ARGV[2]

            local current_size = redis.call('LLEN', queue)
            if current_size >= max_size then
                redis.call('RPOP', queue)
            end
            redis.call('LPUSH', queue, log_entry)
            return current_size + 1
        `;

        this.redisClient.on('error', (err) => {
            if (this._redisAvailable) {
                this._logFallbackWarning(`Redis connection error: ${err.message}. Falling back to local logging.`);
            }
            this._redisAvailable = false;
        });

        this.redisClient.on('ready', () => {
            if (!this._redisAvailable) {
                console.log(`[ICCMLogger INFO] ${this.component}: Successfully connected to Redis.`);
            }
            this._redisAvailable = true;
        });

        // Fire-and-forget connection attempt
        this.redisClient.connect().catch(err => {
            this._logFallbackWarning(`Could not connect to Redis at ${redisUrl}. Reason: ${err.message}. Falling back to local logging.`);
        });
    }

    /** REQ-LIB-005: Dynamically change the log level */
    setLevel(level) {
        const newLevel = LEVELS[level.toUpperCase()];
        if (newLevel !== undefined) {
            this.level = newLevel;
        } else {
            this._logFallbackWarning(`Invalid log level '${level}'. Level not changed.`);
        }
    }

    /** REQ-SEC-001/002/003: Recursively redact sensitive fields */
    _redact(data) {
        if (data === null || typeof data !== 'object') {
            return data;
        }

        if (Array.isArray(data)) {
            return data.map(item => this._redact(item));
        }

        const cleanData = {};
        for (const key in data) {
            if (Object.prototype.hasOwnProperty.call(data, key)) {
                if (this.redactFields.has(key.toLowerCase())) {
                    cleanData[key] = REDACTED_VALUE;
                } else {
                    cleanData[key] = this._redact(data[key]);
                }
            }
        }
        return cleanData;
    }

    /** Internal log method */
    async _log(levelName, message, data, traceId) {
        const levelNum = LEVELS[levelName];
        if (levelNum < this.level) {
            return; // REQ-LIB-006: Filter logs below current level
        }

        const logEntry = {
            component: this.component,
            level: levelName,
            message: message,
            trace_id: traceId || null, // REQ-COR-004, REQ-COR-005
            data: data ? this._redact(data) : null,
            created_at: new Date().toISOString()
        };

        // REQ-LIB-003: Non-blocking push with local fallback
        if (!this._redisAvailable) {
            this._logLocal(logEntry);
            return;
        }

        try {
            // REQ-LIB-004: Atomic FIFO queue management
            await this.redisClient.eval(
                this.luaScript,
                {
                    keys: [this.logQueueName],
                    arguments: [String(this.maxQueueSize), JSON.stringify(logEntry)]
                }
            );
        } catch (err) {
            this._redisAvailable = false;
            this._logFallbackWarning(`Redis command failed. Reason: ${err.message}. Falling back to local logging.`);
            this._logLocal(logEntry);
        }
    }

    /** REQ-LIB-007: Log a warning indicating fallback mode */
    _logFallbackWarning(message) {
        console.warn(`[ICCMLogger WARNING] ${this.component}: ${message}`);
    }

    /** Logs to stdout as a fallback */
    _logLocal(logEntry) {
        console.log(JSON.stringify(logEntry));
    }

    // REQ-LIB-001: Public log methods
    trace(message, data, traceId) { this._log('TRACE', message, data, traceId); }
    debug(message, data, traceId) { this._log('DEBUG', message, data, traceId); }
    info(message, data, traceId) { this._log('INFO', message, data, traceId); }
    warn(message, data, traceId) { this._log('WARN', message, data, traceId); }
    error(message, data, traceId) { this._log('ERROR', message, data, traceId); }

    /** Gracefully close the Redis connection */
    async close() {
        if (this.redisClient && this.redisClient.isOpen) {
            await this.redisClient.quit();
        }
    }
}
